File: 1345_Jump_Game_IV.py
class Solution:
    def minJumps(self, arr: List[int]) -> int:
        n = len(arr)
        nei = collections.defaultdict(list)
        for i, v in enumerate(arr):
            nei[v].append(i)
        queue, visited_num, visited_pos = collections.deque([(0, 0)]), set(), set()
        while queue:
            pos, step = queue.pop()
            if pos == n - 1: return step
            num = arr[pos]
            visited_pos.add(pos)
            for nxt in [pos - 1, pos + 1] + nei[num] * (num not in visited_num):
                if nxt not in visited_pos and 0 <= nxt < n:
                    queue.appendleft((nxt, step + 1))
            visited_num.add(num)
        return 0

File: 1345_Jump_Game_IV.cpp
class Solution {
public:
    int minJumps(vector<int>& arr) {
        int n = arr.size();
        unordered_map<int, vector<int>> graph;
        vector<bool> visited (n, false);
        queue<int> q;
        int steps = 0;
        for (int i = 0; i < n; ++i)
            graph[arr[i]].push_back(i);
        
        visited[0] = true;
        q.push(0);
        while (!q.empty()) {
            for (int i = q.size(); i > 0; --i){
                int t = q.front();
                q.pop();
                if (t == n - 1) return steps;
                vector<int>& next = graph[arr[t]];
                next.push_back(t - 1);
                next.push_back(t + 1);
                for (auto j : next) {
                    if (j >= 0 && j < n && !visited[j]) {
                        visited[j] = true;
                        q.push(j);
                    }
                }
                next.clear();
            }
            steps++;
        }
        
        return 0;
    }
};
